README.md
# Fishy Online
A competetive online multiplayer aquatic life simulator built using HTML5 and websockets.

# Controls
WASD or Arrow keys to move  
Space to activate abilities (if available)  
Escape to focus on the game screen  

# Gameplay
The goal is to become the largest creature in the game. This can be achieved by eating fish smaller than you or eating the food pellets that rain down.

# Creatures
There are a few types of aquatic creatures you can play as.

## Fish
Quick, nimble, and great for collecting food swiftly. Upon eating another creature, fish will drop some uneaten food, which is only edible to bottom feeders.

## Crab
This nifty bottom feeder can swiftly collect scraps raining down from the fish above. Crabs cannot swim too far above the ground, keeping them safe from larger fish above.

## Jellyfish
Jellyfish are significantly slower than fish or crabs, making it harder for them to escape fast preditors. When near the ground, jellyfish can transform into polyps, rooting and making them immune to being eaten. Polyps can still grow from consuming any falling food.

## Clam
The humble clam provides you with a relaxing experience to chill and chat with friends. Clams cannot move nor have any special abilities, but they are completely immune to being eaten. They still can consume any falling food to increase in size.

CHANGELOG.md
# Changelog  
All notable changes to this project will be documented in this file.  
  
## [Unreleased]  

## [0.1.5] - 2023-05-25
### Fixed
- updated socket.io to 4.6.1  

## [0.1.4] - 2022-11-23
### Fixed
- updated dependencies to update nested engine.io package to 6.2.1  

## [0.1.3] - 2022-09-26
### Changed  
- refactored flakes to match enemies operations  
- players now spawn at the top of their `spawnTop`  
  - the top 90px of the screen is now a safe zone  
- set max enemies to 35  
- updated socket.io to version 4.5.2  

### Fixed  
- server spamming `playerScore` events when player intersects larger enemy  
- send score with `playerDeath` event  
- if multiple people disconnect, `removePlayer()` could remove all enemies from the game (again)  

## [0.1.2] - 2022-08-13  
### Added  
- ping display  
- creatures of the same color cannot consume each other  
- new sprites for jellyfish  
- new sprites for blue and red fish  
- spacebar tracking in `wasd`  
- new `playerDeath` event  
  - resets player velocity and jellyfish back to polyps  
- new epic fish names  

### Changed  
- moved supporting server files to `src/`  
- separated ticks to manage:  
  - networking  
  - collision  
  - rendering  
  - second (animation/ping/etc)  
- refactored networking for flakes and enemies  
  - client will render them based position and speed  
  - server will only send on relevant update events  
  - from +30 calls/second to 1-4 calls/second for each  
- jellyfish movement is now more fluid, with no automatic velocity degradation  

### Fixed  
- having a name length of `playerVars.maxNameLength` causes the game not to start  
- if multiple people disconnect, `removePlayer()` could remove all enemies from the game  

## [0.1.1] - 2022-08-01  
### Added  
- CHANGELOG.md  
- FUTURE.md  
  - list of ideas for the future of this game  

### Changed  
- updated express to version 4.18.1  
- updated socket.io to version 4.5.1  
- reduced emitting `enemies` from every tick to every (`tickRate * enemyTickMax`) tick (currently every 10th tick)  
  - reduces socket chatter  
  - client now increments enemy speed to accurately reflect position  
- synchronized tick time for server and client  
- renamed `disconnect` event to `pdisconnect` to avoid socket.io event override error  
- updated `gameloop` player iteration to use new `io.fetchSockets()` promise  
- replaced deprecated 'e.code' with `e.keyCode` for more consistent keybindings  

### Fixed  
- increased fish spawn distance offset with `stageVars.spawnWidthOffset`  
  - reduces jarring spawn of faster fish  

## [0.1.0] - 2020-09  
### Added  
- intial commit of the game  

&nbsp;

src/game.js
const helpers = require('./helpers.js');
const fishNames = require('./fishNames.js');
//const { v4: uuidv4 } = require('uuid');

var isStarted = false;
var reset = false;
var io = null;
var players = {};
var enemies = [];
var flakes = [];

var movementQueue = {};


var startTime, now, then, elapsed;
then = Date.now();
startTime = then;
var tickRate = 1000 / 30;

var stageVars = {
    width: 900,
    height: 700,
    spawnWidthOffset: 10,
    end: {
        roundOver: false,
        winScore: 200,
        countdownMax: 10,
        countdown: 10,
        tickTime: then
    }
};

var playerVars = {
    maxNameLength: 12,
    isLeft: false,
    isPolyp: true,
    collisionWait: 3000,
    fish: {
        sizeX: 16,
        sizeY: 8,
        minSizeX: 10,
        maxSizeX: 150,
        scale: 0.5,
        spawnTop: 90,
        spawnBottom: 525,
        minSpeed: 1,
	    maxSpeed: 7
    },
    crab: {
        sizeX: 16,
        sizeY: 16,
        minSizeX: 16,
        maxSizeX: 64,
        scale: 1,
        spawnTop: 525,
        spawnBottom: 700,
        minSpeed: 1,
        maxSpeed: 2
    },
    clam: {
        sizeX: 16,
        sizeY: 16,
        minSizeX: 16,
        maxSizeX: 128,
        scale: 1,
        spawnTop: 525,
        spawnBottom: 700
    },
    jelly: {
        sizeX: 10,
        sizeY: 16,
        minSizeX: 10,
        maxSizeX: 100,
        scale: 1.5,
        minSpeed: 1,
        maxSpeed: 1,
        spawnTop: 550,
        spawnBottom: 700
    }
};

var enemyVars = {
    minNum: 10,
    maxNum: 35,
    reduceTo: 10,
    incNum: 5,
	minX: 10,
    maxX: 150,
    rareMinX: 300,
    rareMaxX: 500,
    rareChance: 0.2,
	minSpeed: 1,
	maxSpeed: 10,
	delay: 50,
    team: 'fish', //'crab'
    points: 2,
    crabSpawnChance: 0.2
};

var flakeVars = {
    maxNum: 10,
    reduceTo: 20,
    flakeYStart: -5,
    size: 5,
    sizeFish: 10,
    speed: 1,
    delay: 1000,
    points: 1,
    pointsFish: 2,
    flakeMaxSizeX: 100,
    flakeMaxSizeY: 100
};

var colors = [
    'orange',
    'blue',
    'green',
    'purple',
    'black',
    'red'
];

var teams = [
    'fish',
    'crab',
    'jelly',
    'clam'
];

////////////////////////////////////////////////////////

function command(cmd) {
    switch(cmd) {
        case 'reset':
            io.emit('chatMessage', {message: 'Resetting please wait..'});
            reset = true;
            stopGameLoop(true);
            break;
        case 'bigfish':
            addEnemy(1, undefined, true);
            io.emit('enemies', enemies);
            break;
        case 'addenemies':
            enemyVars.reduceTo += 5;
            addEnemy(5);
            io.emit('enemies', enemies);
            break;
        case 'removeenemies':
            enemyVars.reduceTo -= 5;
            break;
        case 'screenclear':
            // TODO: refactor
            enemies = [];
            enemyVars.reduceTo = enemyVars.minNum;
            addEnemy(5);
            flakes = [];
            flakeVars.reduceTo = flakeVars.maxNum;
            addFlake(20);
            Object.keys(players).forEach(function(p) {
                if(enemies.length < enemyVars.maxNum) {
                    addEnemy(5);
                    enemyVars.reduceTo += 5;
                }
            });
            io.emit('enemies', enemies);
            break;
        case 'fishfood':
            flakeVars.reduceTo += 10;
            addFlake(10);
            break;
        default:
            io.emit('chatMessage', {message: 'No command found'});
            break;
    }
}

function validatePlayer(player) {
    // player info validation :^)
    var isValid = true;

    if(typeof player.name === 'undefined' || player.name === '' || player.name.length > playerVars.maxNameLength) {
        isValid = false;
    }
    if(typeof player.team === 'undefined' || player.team === '' || teams.indexOf(player.team) === -1) {
        isValid = false;
    }
    if(typeof player.color === 'undefined' || player.color === '' || colors.indexOf(player.color) === -1) {
        isValid = false;
    }

    return isValid;
}

function addPlayer(id, name, team, color) {
    var newPlayer = {
        name: name.slice(0,playerVars.maxNameLength),
        color: color,
        team: team,
        sizeX: playerVars[team].sizeX,
        sizeY: playerVars[team].sizeY,
        x: Math.floor(Math.random() * Math.floor(stageVars.width/2)) + Math.floor(stageVars.width/4),
        //y: Math.floor((Math.random() * (playerVars[team].spawnBottom - playerVars[team].spawnTop)) + playerVars[team].spawnTop),
        y: playerVars[team].spawnTop - playerVars[team].sizeY,
        isLeft: playerVars.isLeft,
        isPolyp: playerVars.isPolyp,
        score: 0,
        lastCollision: Date.now() - playerVars.collisionWait
    };
    players[id] = newPlayer;
    
    if(enemies.length < enemyVars.maxNum) {
        addEnemy(5);
        enemyVars.reduceTo += 5;
    }
    io.emit('enemies', enemies);
    sendThinFlakes();
}

function removePlayer(id) {
    enemyVars.reduceTo -= 5;
    if(enemies.reduceTo < enemyVars.minNum) {
        enemyVars.reduceTo = enemyVars.minNum;
    }
    if(typeof players[id] !== 'undefined') {
        delete players[id];
    }
}


function addEnemy(qty, index, spawnRare) {
    // remove enemy if reduceTo < enemies.length
    if(typeof index !== 'undefined' && enemyVars.reduceTo < enemies.length) {
        enemies.splice(index, 1);
        io.emit('enemies', enemies);
        return;
    }

    var addQty = 1;
    if(typeof qty !== 'undefined') {
        addQty = qty;
    }

    for(var i = 0; i < addQty; i++) {
        
        var enemy = {};
        if(Math.random() < enemyVars.crabSpawnChance) {
            enemy.team = 'crab';
        }else {
            enemy.team = 'fish';
        }

        enemy.color = colors[Math.floor(Math.random() * colors.length)];
        
        // RARE BOI
        if((typeof spawnRare !== 'undefined' && spawnRare) || (enemy.sizeX === enemyVars.maxX && Math.random() < enemyVars.rareChance)) {
            enemy.team = 'fish';
            enemy.sizeX = Math.floor((Math.random() * enemyVars.rareMaxX) + enemyVars.rareMinX);
            enemy.sizeY = Math.floor(enemy.sizeX/2);
            enemy.speed = Math.random() + 0.5;
            enemy.name = fishNames[Math.floor(Math.random() * fishNames.length)];
        }else { // common enemies
            enemy.sizeX = Math.floor((Math.random() * playerVars[enemy.team].maxSizeX) + playerVars[enemy.team].minSizeX);
            enemy.sizeY = Math.floor(enemy.sizeX * playerVars[enemy.team].scale);
            enemy.speed = (Math.random() * playerVars[enemy.team].maxSpeed) + playerVars[enemy.team].minSpeed;
            enemy.name = enemy.color + ' ' + enemy.team;
        }
        
        // determine direction
        if(Math.floor(Math.random() * 2)) {
            enemy.x = (enemy.sizeX + (enemy.speed * stageVars.spawnWidthOffset)) * -1;
        }else {
            enemy.x = (enemy.sizeX + (enemy.speed * stageVars.spawnWidthOffset)) + stageVars.width;
            enemy.speed *= -1;
        }
        enemy.y = Math.floor((Math.random() * (playerVars[enemy.team].spawnBottom - playerVars[enemy.team].spawnTop)) + playerVars[enemy.team].spawnTop);

        enemy.delay = Math.floor(Math.random() * enemyVars.delay);


        if(addQty === 1 && typeof index !== 'undefined') {
            enemies[index].x = enemy.x;
            enemies[index].y = enemy.y;
            enemies[index].sizeX = enemy.sizeX;
            enemies[index].sizeY = enemy.sizeY;
            enemies[index].speed = enemy.speed;
            enemies[index].team = enemy.team;
            enemies[index].delay = enemy.delay;
            enemies[index].name = enemy.name;
            enemies[index].color = enemy.color;
            sendSingleEnemy(index);
        }else {
            enemies.push(helpers.clone(enemy));
        }
    }
}

function addFlake(qty, index, location) {
    // remove flake if reduceTo < flakes.length
    if(typeof index !== 'undefined' && flakeVars.reduceTo < flakes.length) {
        flakes.splice(index, 1);
        sendThinFlakes();
        return;
    }

    var addQty = 1;
    if(typeof qty !== 'undefined') {
        addQty = qty;
    }

    for(var i = 0; i < addQty; i++) {
        
        var flake = {};
        if(typeof location !== 'undefined') {
            flake.x = location.x;
            flake.y = location.y;
            flake.sizeX = flakeVars.sizeFish;
            flake.sizeY = flakeVars.sizeFish;
            flake.delay = 0;
            flake.type = 'fish';
            flake.points = flakeVars.pointsFish;
        }else {
            flake.x = Math.floor((Math.random() * stageVars.width));
            flake.y = flakeVars.flakeYStart;
            flake.sizeX = flakeVars.size;
            flake.sizeY = flakeVars.size;
            flake.delay = Math.floor(Math.random() * flakeVars.delay);
            flake.type = '';
            flake.points = flakeVars.points;
        }
        //flake.speed = flakeVars.speed;

        if(addQty === 1 && typeof index !== 'undefined') {
            flakes[index].x = flake.x;
            flakes[index].y = flake.y;
            flakes[index].sizeX = flake.sizeX;
            flakes[index].sizeY = flake.sizeY;
            flakes[index].delay = flake.delay;
            flakes[index].type = '';
            flakes[index].points = flake.points;
            sendSingleFlake(i);
        }else {
            flakes.push(helpers.clone(flake));
        }
    }
}

//////////////////////////////////////////////////////////////////////////////////////////

function startGameLoop(myIO) {
    if(!isStarted) {
        if(typeof myIO !== 'undefined') {
            io = myIO;
        }
        isStarted = true;
        reset = false;

        enemyVars.reduceTo = enemyVars.minNum;
        flakeVars.reduceTo = 10;

        // init enemies
        addEnemy(5);

        Object.keys(players).forEach(function(p) {
            if(enemies.length < enemyVars.maxNum) {
                addEnemy(5);
                enemyVars.reduceTo += 5;
            }
        });
        io.emit('enemies', enemies);

        // init flakes
        addFlake(20);
        sendThinFlakes();

        gameloop();
    }
}

function stopGameLoop(isReset) {
    isStarted = false;
    enemies = [];
    flakes = [];

    // TODO: reset players (score + location)
    Object.keys(players).forEach(function(p) {
        players[p].sizeX = playerVars[players[p].team].sizeX;
        players[p].sizeY = playerVars[players[p].team].sizeY;
        players[p].score = 0;
        players[p].isPolyp = true;
        players[p].lastCollision = Date.now() - playerVars.collisionWait;
    });
    io.emit('players', players);
    io.emit('enemies', enemies);
    sendThinFlakes();

    if(typeof isReset !== 'undefined' && isReset) {
        setTimeout(function() {
            io.emit('chatMessage', {message: 'Reset complete!'});
            startGameLoop();
        }, 500);
    }
}

//////////////////////////////////////////////////////////////////////////////////////////

function collision() {
    // TODO: validate player collision

    var rightNow = Date.now();

    // enemies
    for(var e = 0; e < enemies.length; e++) {
        var ce = enemies[e];
        if(ce.delay > 0) {
            ce.delay--;
            if(ce.delay == 0) {
                sendSingleEnemy(e);
            }
        }else {
            // move enemy
            ce.x += ce.speed;
            // if OOB
            if((ce.speed > 0 && ce.x - ce.sizeX > stageVars.width) ||
                (ce.speed < 0 && ce.x + ce.sizeX < 0)) {
                    addEnemy(1, e);
                    continue;
            }
            // if collision with enemy
            Object.keys(players).forEach(function(p) {
                // ignore if clam/polyp
                if(players[p].team === 'clam' || (players[p].team === 'jelly' && players[p].isPolyp)) {
                    return;
                }
                // ignore enemies on your team
                if(players[p].color == ce.color) {
                    return;
                }
                // ignore if player had recent collision action
                if(players[p].lastCollision + playerVars.collisionWait >= rightNow) {
                    return;
                }
                
                // if initial intersection
                if(helpers.intersect(ce, players[p])) {

                    // player is larger, consume enemy
                    if((players[p].team === 'jelly' && players[p].sizeY > ce.sizeX) || (players[p].team !== 'jelly' && players[p].sizeX > ce.sizeX)) {
                        if(players[p].sizeX < playerVars[players[p].team].maxSizeX) {
                            if(players[p].sizeX < playerVars[players[p].team].maxSizeX) {
                                players[p].sizeX++;
                                players[p].sizeY = Math.floor(players[p].sizeX * playerVars[players[p].team].scale);
                            }
                        }
                        players[p].score += enemyVars.points;
                        addEnemy(1, e);

                        addFishFlake(p, false);
                        sendThinPlayerScore(p);

                        checkWinCondition(p);
                    // enemy is larger - deduct points, reset position, broadcast
                    }else {
                        if(ce.team == "fish") {
                            if(helpers.intersectFishPlayer(ce, players[p])) {
                                playerDeathReset(p, ce.name, ce.sizeX, rightNow);
                            }
                        }else {
                            playerDeathReset(p, ce.name, ce.sizeX, rightNow);
                        }
                    }
                }
            });

            // TODO: optimize; causes stuttering
            /*
            for(var ee = 0; ee < enemies.length; ee++) {
                if(e != ee) {
                    var cee = enemies[ee];
                    if(ce.color != cee.color && helpers.intersect(ce, cee)) {
                        if(ce.sizeX > cee.sizeX) {
                            addEnemy(1, ee);
                            addFishFlake(ce, false);
                            io.emit('enemies', enemies);
                            enemyTick = 0;
                            break;
                        }
                    }
                }
            }
            */
        }
    }//for enemies

    // flakes
    for(var f = 0; f < flakes.length; f++) {
        var cf = flakes[f];
        if(cf.delay > 0) {
            cf.delay--;
            if(cf.delay == 0) {
                sendSingleFlake(f);
            }
        }else {
            // move flake
            cf.y += flakeVars.speed;
            // if OOB
            if(cf.y > stageVars.height) {
                addFlake(1, f);
                sendSingleFlake(f);
                continue;
            }
            // if collision with enemy
            Object.keys(players).forEach(function(p) {
                if(typeof flakes[f] === 'undefined') {
                    return;
                }
                if(players[p].team === flakes[f].type) {
                    return;
                }
                if(players[p].team !== 'clam' && (players[p].sizeY >= flakeVars.flakeMaxSizeY || players[p].sizeX >= flakeVars.flakeMaxSizeX)) {
                    return;
                }
                if((players[p].lastCollision + playerVars.collisionWait < rightNow) && 
                    helpers.intersect(cf, players[p])) {

                        if(players[p].sizeX < playerVars[players[p].team].maxSizeX) {
                            players[p].sizeX++;
                            players[p].sizeY = Math.floor(players[p].sizeX * playerVars[players[p].team].scale);
                        }
                        players[p].score += flakes[f].points;

                        // send flakes
                        addFlake(1, f);
                        sendSingleFlake(f);

                        sendThinPlayerScore(p);
                        checkWinCondition(p);
                        return;
                }
            });
        }
    }//for flakes
}//collision

function playerDeathReset(playerId, killerName, killerSize, rightNowTime) {
    players[playerId].x = Math.floor(stageVars.width/2 - (players[playerId].sizeX/2));
    //players[playerId].y = stageVars.height - players[playerId].sizeY - 10;
    players[playerId].y = playerVars[players[playerId].team].spawnTop - players[playerId].sizeY;
    players[playerId].lastCollision = rightNowTime;
    players[playerId].score -= enemyVars.points;
    // player death time + collision wait = client to show invulnerable animation
    players[playerId].diedAt = rightNowTime + playerVars.collisionWait;

    var thinPlayerMove = {
        x: players[playerId].x,
        y: players[playerId].y,
        isLeft: players[playerId].isLeft,
        diedAt: players[playerId].diedAt
    };
    addMovement(playerId, thinPlayerMove);

    var resp = {};
    resp.id = playerId;
    resp.message = players[playerId].name + '[' + (players[playerId].team === 'jelly' ? players[playerId].sizeY : players[playerId].sizeX) + '] was eaten by ' + killerName + '[' + killerSize + ']';
    resp.score = players[playerId].score;
    io.emit('playerDeath', resp);
}//playerDeathReset

// use combination of setImmediate delay + delta time for smoothness
function gameloop() {
    io.fetchSockets()
      .then((sockets) => {
        if(sockets.length > 0 && !reset) {
            now = Date.now();
            elapsed = now - then;
            if(elapsed > tickRate) {
                then = now - (elapsed % tickRate);
    
                collision();
                movement();
                if(stageVars.end.roundOver) {
                    roundOver(now);
                }
            }
            setImmediate(gameloop, tickRate);
        }else {
            stopGameLoop();
        }
      })
      .catch(console.log);
}

function addMovement(id, data) {
    if(movementQueue[id] === undefined) {
        movementQueue[id] = {};
    }
    movementQueue[id].x = data.x;
    movementQueue[id].y = data.y;
    movementQueue[id].isLeft = data.isLeft;
    if(typeof data.diedAt !== 'undefined') {
        movementQueue[id].diedAt = data.diedAt;
    }
}
function movement() {
    if(Object.keys(movementQueue).length > 0) {
        io.emit('playersMoved', movementQueue);
        movementQueue = {};
    }
}

function addFishFlake(id, reduceSize) {
    if(typeof players[id] !== 'undefined' && players[id].team === 'fish' && players[id].sizeX > 1) {
        var location = {
            x: players[id].x + (players[id].sizeX / 2),
            y: players[id].y + (players[id].sizeY / 2)
        };
        addFlake(1, undefined, location);
        sendThinFlakes();

        if(typeof reduceSize !== 'undefined' && reduceSize) {
            players[id].sizeX--;
            players[id].sizeY = Math.floor(players[id].sizeX * playerVars[players[id].team].scale);
            
            sendThinPlayerScore(id);
        }
    }
}

function checkWinCondition(id) {
    if(stageVars.end.roundOver || typeof players[id] === 'undefined') {
        return;
    }
    if(players[id].score >= stageVars.end.winScore) {
        stageVars.end.roundOver = true;
        io.emit('chatMessage', {message: players[id].name + '[' + players[id].team + '] has won the round!'});
    }
}
function roundOver(now) {
    if(stageVars.end.tickTime < now) {
        stageVars.end.countdown--;
        if(stageVars.end.countdown < 0) {
            stageVars.end.countdown = stageVars.end.countdownMax;
            stageVars.end.roundOver = false;
            io.emit('chatMessage', {message: 'Resetting please wait..'});
            reset = true;
            stopGameLoop(true);
        }else {
            io.emit('chatMessage', {message: 'Resetting in ' + stageVars.end.countdown});
            stageVars.end.tickTime = now + 1000;
        }
    }
}


//#region send socket events
function sendSingleEnemy(index) {
    io.emit('enemy', {
        id: index,
        enemy: enemies[index]
    });
}
function sendThinFlakes() {
    var tf = [];
    flakes.forEach(function(f) {
        tf.push({
            x: f.x,
            y: f.y,
            size: f.sizeX,
            delay: f.delay
        });
    });
    io.emit('flakes', tf);
}
function sendSingleFlake(index) {
    // TODO: investigate race condition
    if(typeof flakes[index] !== "undefined") {
        io.emit('flake', {
            id: index,
            flake: {
                x: flakes[index].x,
                y: flakes[index].y,
                size: flakes[index].sizeX,
                delay: flakes[index].delay
            }
        });
    }
}
function sendThinPlayerScore(index) {
    io.emit('playerScore', {
        id: index,
        sizeX: players[index].sizeX,
        sizeY: players[index].sizeY,
        score: players[index].score
    });
}
//#endregion

module.exports = {
    players,
    startGameLoop,
    stopGameLoop,
    validatePlayer,
    addPlayer,
    removePlayer,
    command,
    addMovement,
    addFishFlake
};